### PR TITLE
Implement CAP-58 (constructor support)

### PR DESCRIPTION
**Pull request by dmkozh**
### What

The implementation itself is fairly simple (thanks to not needing to support v21 semantics) and follows the [CAP-58](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0058.md). It only has a few caveats:

- Pre-v22 protocol Wasms are treated as if they don't have a constructor (they shouldn't have it anyway, but just in case they do we ignore it in order to not break the contract semantics). Thus we don't need to even try calling constructor for them
- We try to call constructor for all v22+ contracts, so we charge for VM instantiation, which increases the creation cost to some degree. I'm not sure if we can discount calling non-existent functions; if we find that necessary, we can still improve this before the protocol release.

Bulk of this change is tests, and testing is really not trivial as we want to cover a set product of several different options:
- p21 vs p22 contracts
- v1 vs v2 host function
- Constructor return values and errors
- constructor arguments present/not present/'default' constructor
- auth for constructor itself and inside the constructor
- deployer support
- invoker auth support
- custom account support

To make things worse, a contract may only have a single constructor, so this needed a lot of new test Wasms.

### Why

See CAP-58 motivation.

### Known limitations

I'm sure I haven't covered every possible combination of parameters in tests, but at least tried to have some basic coverage for most of the cases and most obvious combinations.

### CODE REVIEW

**Comment by anupsdf** [comment]
If a pre-v22 contract gets its wasm updated in v22+ protocol with `update_current_contract_wasm` host fn call and the new wasm has a constructor, we will not be calling the constructor right?

**Comment by anupsdf** [comment]
> If a pre-v22 contract gets its wasm updated in v22+ protocol with `update_current_contract_wasm` host fn call and the new wasm has a constructor, we will not be calling the constructor right?

Nevermind, I see this mentioned in the [CAP](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0058.md#interaction-with-contract-updates), 
`When the contract has its code updated it is not considered created and thus constructor won't be called.`

**Review by sisuresh — COMMENTED** [review]
Just a couple comments. I still need to look at the auth changes.

File: soroban-env-host/src/vm.rs [hunk 1]
```diff
@@ -402,12 +403,16 @@ impl Vm {
...
         {
             None => {
-                return Err(host.err(
-                    ScErrorType::WasmVm,
-                    ScErrorCode::MissingValue,
-                    "invoking unknown export",
-                    &[func_sym.to_val()],
-                ))
+                if treat_missing_function_as_noop {
+                    return Ok(Val::VOID.into());
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/vm.rs:407]: Conceptually it's the cost of calling the contract, so it's not really significant IMO. Majority of interactions do at least one contract call, so the cost of creating the contract won't be significantly different from what's already happening most of the time. In absolute terms, it's a bit hard to tell as it depends on the contract, but in the updated tests I can see +~300k-+2M instructions, which I think isn't really worth bothering. That said,  I want to examine the simulation observation shifts regardless, as the increase is suspiciously high there. But I don't think that's a good enough reason to come up with workarounds unless they're trivial. I'm not even sure we can save that much as we still need to figure out if the function is there at all.

File: soroban-env-host/src/auth.rs [hunk 2]
```diff
@@ -602,7 +602,10 @@ impl AuthorizedFunction {
                     args: host.scvals_to_val_vec(xdr_contract_fn.args.as_slice())?,
                 })
             }
-            SorobanAuthorizedFunction::CreateContractHostFn(xdr_args) => {
+            SorobanAuthorizedFunction::CreateContractHostFn(_) => {
+                return Err(host.err(ScErrorType::Auth, ScErrorCode::InvalidInput, "SorobanAuthorizedFunction::CreateContractHostFn is deprecated in authorization payloads. Please use SorobanAuthorizedFunction::CreateContractHostFnV2 instead.", &[]));
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/auth.rs]: Good point, updated.

File: soroban-env-host/src/vm.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/vm.rs:407]: Actually, 2M instructions number is not realistic, we should probably be below 1M instructions. We should be caching VM for all the Wasms in the footprint anyways, so we're adding a call to the cached VM, which should be cheap.
Also for the simulation diff I forgot that we're using arbitrary costs (just to make the test independent of the actual costs as this library takes the costs as the external input), so +2M instructions added there doesn't mean anything.

File: soroban-simulation/src/simulation.rs [hunk 3]
```diff
@@ -141,6 +141,7 @@ pub fn simulate_invoke_host_function_op(
         base_prng_seed,
         &mut diagnostic_events,
     );
+    dbg!(&budget);
```
Comment (sisuresh) [inline soroban-simulation/src/simulation.rs:144]: Did you mean to include this?

Comment (dmkozh) [reply soroban-simulation/src/simulation.rs]: Not really, removed.

**Review by sisuresh — APPROVED** [review]

**Pull request merged**